crates/gitbutler-repo/tests/change_reference.rs

```diff
@@ -0,0 +1,254 @@
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/asdf".into(),
+        wrong_change.clone(),
+    );
+    assert!(result.is_err());
+    Ok(())
+}
+
+#[test]
+fn create_fails_when_change_id_not_found() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    let result = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/baz".into(),
+        "does-not-exist".into(),
+    );
+    assert!(result.is_err());
+    Ok(())
+}
+
+#[test]
+fn list_success() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    let first_ref = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/first".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    )?;
+    let second_ref = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/second".into(),
+        test_ctx.commits.last().unwrap().change_id().unwrap(),
+    )?;
+    let result = list_branch_references(&ctx, test_ctx.branch.id)?;
+    assert_eq!(result.len(), 2);
+    assert_eq!(result[0], first_ref);
+    assert_eq!(result[1], second_ref);
+    Ok(())
+}
+
+#[test]
+fn update_success() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/first".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    )?;
+    let updated = update_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/first".into(),
+        test_ctx.commits.last().unwrap().change_id().unwrap(),
+    )?;
+    assert_eq!(
+        updated.change_id,
+        test_ctx.commits.last().unwrap().change_id().unwrap()
+    );
+    let list = list_branch_references(&ctx, test_ctx.branch.id)?;
+    assert_eq!(list.len(), 1);
+    assert_eq!(
+        list[0].change_id,
+        test_ctx.commits.last().unwrap().change_id().unwrap()
+    );
+    Ok(())
+}
+
+#[test]
+fn push_success() -> Result<()> {
+    let (ctx, _temp_dir) = command_ctx("multiple-commits")?;
+    let test_ctx = test_ctx(&ctx)?;
+    let reference = create_change_reference(
+        &ctx,
+        test_ctx.branch.id,
+        "refs/remotes/origin/first".into(),
+        test_ctx.commits.first().unwrap().change_id().unwrap(),
+    )?;
+    let result = push_change_reference(
+        &ctx,
+        reference.branch_id,
+        reference.name,
+        false,
+        &Helper::default(),
+    );
+    assert!(result.is_ok());
+    Ok(())
+}
+
+fn command_ctx(name: &str) -> Result<(CommandContext, TempDir)> {
+    gitbutler_testsupport::writable::fixture("stacking.sh", name)
```

Comment:
I wonder if it's easier or faster to use this over, day, using the `git2` based test-project initialization.
Never tested it.

In theory, this can be faster as it will just copy files around, even though this is configured to execute the script which then might be slower thanks to the program invocation overhead.

Or in other words: even though I am using these for read-only fixtures, for writable fixtures using the standard test-support might be faster.

Reply:
i see. i kinda wanted to test this out, and ended getting myself into the situation of needing to expose the scripted fixtures in a writable form. 
Having to write bash can be annoying, so perhaps the other approach is also more ergonomic

Reply:
Yes, `bash` is great only when there are no nicer APIs around, and it's really nice for read-only fixtures with unbeatable performance.
But for everything writable, it's probably not very desirable unless the `Copy` mode can be used so the script isn't executed every time.